frameworks/PHP/symfony/src/Controller/DbController.php

```diff
@@ -10,66 +9,74 @@
 
 class DbController
 {
+
     /** @var EntityManagerInterface */
     private $entityManager;
+
     /** @var WorldRepository */
     private $worldRepository;
 
+    private function getUniqueRandomNumbers($count, $min, $max)
+    {
+        $res = array();
+        do {
+            $res[\mt_rand($min, $max)] = 1;
+        } while (\count($res) < $count);
+        return \array_keys($res);
+    }
+
     public function __construct(EntityManagerInterface $entityManager, WorldRepository $worldRepository)
     {
         $this->entityManager = $entityManager;
         $this->worldRepository = $worldRepository;
     }
 
     /**
+     *
      * @Route("/db")
      */
     public function db(): JsonResponse
     {
-        $world = $this->worldRepository->find(mt_rand(1, 10000));
-
-        return new JsonResponse($world);
+        return new JsonResponse($this->worldRepository->find(\mt_rand(1, 10000)));
     }
 
     /**
+     *
      * @Route("/queries")
      */
     public function queries(Request $request): JsonResponse
     {
         $queries = (int) $request->query->get('queries', 1);
-        $queries = min(max($queries, 1), 500);
+        $queries = \min(\max($queries, 1), 500);
 
         // possibility for enhancement is the use of SplFixedArray -> http://php.net/manual/de/class.splfixedarray.php
         $worlds = [];
-
-        for ($i = 0; $i < $queries; ++$i) {
-            $worlds[] = $this->worldRepository->find(mt_rand(1, 10000));
+        $numbers = $this->getUniqueRandomNumbers($queries, 1, 10000);
+        foreach ($numbers as $id) {
+            $worlds[] = $this->worldRepository->find($id);
         }
 
         return new JsonResponse($worlds);
     }
 
     /**
+     *
      * @Route("/updates")
      */
     public function update(Request $request): JsonResponse
     {
         $queries = (int) $request->query->get('queries', 1);
-        $queries = min(500, max(1, $queries));
+        $queries = \min(500, \max(1, $queries));
 
         $worlds = [];
 
-        for ($i = 0; $i < $queries; ++$i) {
-            $world = $this->worldRepository->find(mt_rand(1, 10000));
-            if ($world) {
-                $randomNumber = mt_rand(1, 10000);
-                $world->setRandomNumber($randomNumber);
-                $worlds[] = $world;
-            }
+        $numbers = $this->getUniqueRandomNumbers($queries, 1, 10000);
+        foreach ($numbers as $id) {
+            $world = $this->worldRepository->find($id);
+            $world->setRandomNumber(\mt_rand(1, 10000));
+            $worlds[] = $world;
+            $this->entityManager->flush();
```

Comment:
It was the only flush for 20 updates that caused failed transactions and missing queries.
I also tried by flushing in groups of 2 instances -> same problems. 

We have to keep in mind the concurrency level (512), and the fact that the database is Mysql.
An update of 20 instances in a single transaction does not pass.

Reply:
did you compare time before/after?.

Whats's about using a retry loop instead?

Reply:
>did you compare time before/after?.

It's obviously faster with just one flush (after the loop).
But the update becomes blocking for a longer duration, so it doesn't support concurrent access. 
>Whats's about using a retry loop instead?

I'll try with a retry, provided the flush gives me that opportunity.

Reply:
I tested this `flush with retry` version:
```php
$co = $this->entityManager->getConnection();
$co->beginTransaction();
$done = false;
do {
    try {
        $this->entityManager->flush();
        $co->commit();
        $done = true;
    } catch (\Exception $e) {
        $co->rollback();
    }
} while (! $done);
```
This is a failure, since in case of an exception, the entityManager closes:
![image](https://user-images.githubusercontent.com/2511052/68697450-19cbe600-057f-11ea-8470-6a03414aa2b4.png)

It is therefore necessary to make a version where, in case of exception, we have to:
- reopen the entityManager
- reattach instances that have not been updated
```php
private function flushUpdates(EntityManagerInterface $em, array $worlds)
{
    $co = $em->getConnection();
    do {
        try {
            $co->beginTransaction();
            $em->flush();
            $co->commit();
            $done = true;
        } catch (\Exception $e) {
            $done = false;
            $co->rollback();
            if (! $em->isOpen()) {
                $em = $em->create($co, $em->getConfiguration());
                foreach ($worlds as $world) {
                    $world = $em->merge($world);
                }
            }
        }
    } while (! $done);
}
```
Performances are significantly better than with flushing on each instance, but as a result, too many db queries are sent.
If you see a better solution.....

Reply:
the begin transaction is not in the loop in the first example

Reply:
> the begin transaction is not in the loop in the first example

Good point, but it's just an error when copying and pasting on github.
This first version does not work

Reply:
I'll try by introducing a batchsize, to limit deadlocks.

Reply:
those deadlock are really a pain.
I've fixed them with
```
$con = $this->entityManager->getConnection();
        $con->beginTransaction();
        foreach ($this->getUniqueRandomNumbers($queries) as $id) {
            $worlds[] = $world = $this->worldRepository->find($id, LockMode::PESSIMISTIC_WRITE);
            do {
                $newId = mt_rand(1, 10000);
            } while($id === $newId);
            $world->setRandomNumber($newId);
        }

        $this->entityManager->flush();
        $con->commit();
```
but... this have a performance impact...

Reply:
I'll try your version....

Reply:
Your version would have been much more elegant, but it doesn't pass the check.
Exception management should be added...
![image](https://user-images.githubusercontent.com/2511052/68768772-71bc2880-0623-11ea-8eac-0d0104db76b0.png)

I think it is better to limit the size of the batch, and do less at the same time, to limit the deadlock risk.
As soon as you have exceptions, retrying is very costly.

Reply:
ok, I've something: Instead of dealing with lock, retry and concurrency. I just changed the test to avoid deadlock (a deadlock occurs when a process 1 update A then B and a process 2 update B then A). For that, I just sort ids.
```php
        $ids = $this->getUniqueRandomNumbers($queries);
        sort($ids);
        foreach ($ids as $id) {
```
I don't have deadlock anymore

Reply:
for the record, I also have to assert that the id is changed. With 512 concurrent connection and 20 update you have 10 240 queries. Given the random is between 1 and 10 000 there is a significant probability to pick the same id => doctrine won't perform the `update` request => Test failure